URL Rewriting Middleware topic updates for 2.0

guardrex opened the pull request:
Fixes #4004 
[Internal review URL](https://review.docs.microsoft.com/en-us/aspnet/core/fundamentals/url-rewriting?branch=pr-en-us-4005&tabs=aspnetcore1x)

dnfclas commented:
@guardrex,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

guardrex commented:
@Rick-Anderson @tdykstra 

cc/ @mikaelm12

Notes:
* Adopted the sample layout spec we discussed yesterday. *I like it!*
* Updated the 1.x sample to latest package versions on 1.x.y.
* Broke some long sample app lines into two lines and shortened some strings to help with viewing content on narrow width devices.
* I'm going with the `WebHost` extension methods in the 2.0 app: I like the way it gets all of the setup bits in one place. :+1: for the 😵 Nutjob:tm: 😵 startup pattern. lol
* Cleaned up the text; the usual stuff ...
   - Addressed guidelines adherence issues (contractions, :hocho: the future tense, etc.)
   - "application" :point_right: "app"
   - Better word and phrase choices
* 🥀 Table Beautification 🌷 I like what we were doing on the .NET Core docs side, so I'm going to start doing that over here. When editing a topic, groking table content is easier when they're formatted in markdown.
* Updated the *Unsupported features* section and added tabs for this content so that it's easy for the reader to see what isn't supported for each major release.

*It's off of WIP and ready for* :eyes:!

I'm moving on to the next one ... either the response compression topic or the response caching one.

Rick-Anderson commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -16,76 +16,92 @@ uid: fundamentals/url-rewriting
 * Preventing image hotlinking
 
-You can define rules for changing the URL in several ways, including regular expression (regex) matching rules, rules based on the Apache mod_rewrite module, rules based on the IIS Rewrite Module, and with your own method and class rule logic. This document introduces URL rewriting with instructions on how to use URL Rewriting Middleware in ASP.NET Core applications.
+You can define rules for changing the URL in several ways, including regular expression (regex) matching rules, rules based on the Apache mod_rewrite module, rules based on the IIS Rewrite Module, and with your own method and class rule logic. This document introduces URL rewriting with instructions on how to use URL Rewriting Middleware in ASP.NET Core apps.
```
TL4MT (Too long for machine translation)
Rules for changing a URL can be defined in several ways:

* regex

Rick-Anderson commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -16,76 +16,92 @@ uid: fundamentals/url-rewriting
+
+---
+
+In a browser with developer tools enabled, make a request to the sample app with the path `/redirect-rule/1234/5678`. The regex matches the request path on `redirect-rule/(.*)`, and the path is replaced with `/redirected/1234/5678`. The redirect URL is sent back to the client with a 302 (Found) status code. The browser makes a new request at the redirect URL, which appears in the browser's addres ...
```
A ~complete~ round trip is made to the server when a URL is *redirected*.

Rick-Anderson commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -16,76 +16,92 @@ uid: fundamentals/url-rewriting
 ![Browser window with Developer Tools tracking the requests and responses](url-rewriting/_static/add_redirect.png)
 
-The part of the expression contained by parentheses is called a *capture group*. The dot (`.`) of the expression means *match any character*, and the asterisk (`*`) signifies to *match the preceding character zero or more times*. Therefore, the last two path segments of the URL, `1234/5678`, are captured by capture group `(.*)`. Any value you provide in the request URL after `redirect-rule/` will ...
+The part of the expression contained within parentheses is called a *capture group*. The dot (`.`) of the expression means *match any character*, and the asterisk (`*`) indicates *match the preceding character zero or more times*. Therefore, the last two path segments of the URL, `1234/5678`, are captured by capture group `(.*)`. Any value you provide in the request URL after `redirect-rule/` is  ...
```
The dot (`.`) of the expression means *match any character*.  The asterisk (`*`) indicates *match the preceding character zero or more times*.

Rick-Anderson commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -16,76 +16,92 @@ uid: fundamentals/url-rewriting
 <a name=url-redirect-to-secure-endpoint></a>
 ### URL redirect to a secure endpoint
-Use `AddRedirectToHttps()` to redirect insecure requests to the same host and path with secure HTTPS protocol (`https://`) with the flexibility to choose the status code and port. If the status code is not supplied, the middleware will default to 302 (Found). If the port is not supplied, the middleware will default to `null`, which means the protocol will change to `https://` and the client will  ...
+Use `AddRedirectToHttps()` to redirect insecure requests to the same host and path with secure HTTPS protocol (`https://`) with the flexibility to choose the status code and port. If the status code is not supplied, the middleware defaults to 302 (Found). If the port is not supplied, the middleware defaults to `null`, which means the protocol changes to `https://` and the client accesses the reso ...
```
Use `AddRedirectToHttps`  - don't  include `()`

Rick-Anderson replied:
+Use `AddRedirectToHttps` to redirect HTTP requests to the same host and path using  HTTPS (`https://`). `AddRedirectToHttps`  can specify the status code and port. 
just trying to split it up.

Rick-Anderson commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -16,76 +16,92 @@ uid: fundamentals/url-rewriting
+Use `AddRedirectToHttpsPermanent()` to redirect insecure requests to the same host and path with secure HTTPS protocol (`https://` on port 443). The middleware sets the status code to 301 (Moved Permanently).
 
-The sample application is capable of demonstrating how to use `AddRedirectToHttps()` or `AddRedirectToHttpsPermanent()`. Add the extension method to the `RewriteOptions()`. Make an insecure request to the application at any URL. In order to see the response in a browser, you will probably need to dismiss a browser security warning that the self-signed certificate is untrusted.
+The sample app is capable of demonstrating how to use `AddRedirectToHttps()` or `AddRedirectToHttpsPermanent()`. Add the extension method to the `RewriteOptions()`. Make an insecure request to the app at any URL. Dismiss the browser security warning that the self-signed certificate is untrusted.
```
Can you do a global search replace and get rid of `()`?

Rick-Anderson commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -96,47 +112,75 @@ Original Request using `AddRedirectToHttpsPermanent()`: `/secure`
 
 > [!NOTE]
-> `skipRemainingRules: true` should be used whenever possible, because matching rules is an expensive process and slows down application response time. For the fastest application response, order your rewrite rules from most frequently matched to least frequently matched and skip the processing of the remaining rules when a match occurs and no additional rule processing is required.
+> Use `skipRemainingRules: true` whenever possible, because matching rules is an expensive process and reduces app response time. For the fastest app response, order your rewrite rules from the most frequently matched rule to the least frequently matched rule and skip the processing of the remaining rules when a match occurs and no additional rule processing is required.
```
For the fastest app response:
* order your rewrite rules from the most frequently matched rule to the least frequently matched rule.
* skip the processing of the remaining rules when a match occurs and no additional rule processing is required.

BTW, we usually end each list element with punctuation.

Rick-Anderson commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -96,47 +112,75 @@ Original Request using `AddRedirectToHttpsPermanent()`: `/secure`
+
+# [ASP.NET Core 1.x](#tab/aspnetcore1x)
+
+The first parameter takes an `IFileProvider`, which is provided in the sample app via [Dependency Injection](dependency-injection.md) by injecting the `IHostingEnvironment` and using it to provide the `ContentRootFileProvider`. The second parameter is the path to your rules file, which is *ApacheModRewrite.txt* in the sample app.
```
The first parameter takes an `IFileProvider`, which is provided in the sample app via [Dependency Injection](dependency-injection.md) by injecting the `IHostingEnvironment` and using it to provide the `ContentRootFileProvider`. 

2 or more sentences

Rick-Anderson commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -175,30 +219,65 @@ The middleware supports the following Apache mod_rewrite server variables:
 
 ### IIS URL Rewrite Module rules
-To use rules that would normally apply to the IIS URL Rewrite Module, use `AddIISUrlRewrite()`. The first parameter takes an `IFileProvider`, while the second parameter is the path to your XML rules file, which is *IISUrlRewrite.xml* in the sample application. You must make sure that the rules file is deployed with the application. Don't point this at your *web.config* file, as these rules should ...
+To use rules that apply to the IIS URL Rewrite Module, use `AddIISUrlRewrite()`. Make sure that the rules file is deployed with the app. Don't direct the middleware to use your *web.config* file when running on Windows Server IIS, as these rules should be stored outside of your *web.config* to avoid conflicts with the IIS Rewrite module. For more information and examples of IIS URL Rewrite Module ...
```
Don't direct the middleware to use your *web.config* file when running on Windows Server IIS. With IIS these rules should be stored outside of your *web.config* to avoid conflicts with the IIS Rewrite module.

Rick-Anderson commented:
@guardrex  fabulous work!

mikaelm12 commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -16,127 +16,173 @@ uid: fundamentals/url-rewriting
 ![A WebAPI service endpoint has been temporarily changed from version 1 (v1) to version 2 (v2) on the server. A client makes a request to the service at the version 1 path /v1/api. The server sends back a 302 (Found) response with the new, temporary path for the service at version 2 /v2/api. The client makes a second request to the service at the redirect URL. The server responds with a 200 (OK)  ...
 
-When redirecting requests to a different URL, you will indicate whether the redirect is permanent or temporary. The 301 (Moved Permanently) status code is used where the resource has a new, permanent URL and you wish to instruct the client that all future requests for the resource should use the new URL. The client will cache the response when a 301 status code is received. The 302 (Found) status ...
+When redirecting requests to a different URL, you indicate whether the redirect is permanent or temporary. The 301 (Moved Permanently) status code is used where the resource has a new, permanent URL and you wish to instruct the client that all future requests for the resource should use the new URL. The client may cache the response when a 301 status code is received. The 302 (Found) status code  ...
```
I tempted to ask you to emphasize that when using a 301 that the browser will cache the response. It's a  potentially frustrating trap that people can fall into easily when redirecting.

guardrex replied:
Sure ... I'm breaking for dinner but will get that in tonight.

mikaelm12 replied:
Cool. No rush 😄

mikaelm12 commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -16,127 +16,173 @@ uid: fundamentals/url-rewriting
-[!code-csharp[Main](url-rewriting/sample/Startup.cs?name=snippet1)]
+# [ASP.NET Core 1.x](#tab/aspnetcore1x)
+
+[!code-csharp[Main](url-rewriting/samples/1.x/Startup.cs?name=snippet1)]
```
I don't think this is rendering the way you intended

guardrex replied:
How do u mean?

guardrex replied:
Seems ok here ... what's it doing there?

Rick-Anderson replied:
I looked at the build and it looked OK.

mikaelm12 replied:
![rewrite-doc-capture](https://user-images.githubusercontent.com/4809660/29438190-e445f230-8369-11e7-96be-f84450c629a3.PNG)
Is it supposed to look like this?

Rick-Anderson replied:
Nope
`[!code-csharp[Main](`

mikaelm12 replied:
What I mean is it just looks a bit odd to me. But if that's how you guys want it that's fine I guess

guardrex replied:
I saw a few authors in a few spots put the name of the class, method, etc. there, but it only renders in the `name` attribute, thus fairly pointless ...
```html
<code class="lang-csharp" data-src="https://github.com/guardrex/Docs/blob/guardrex/url-
rewriting-middleware-2-0/aspnetcore/fundamentals/url-rewriting/samples/1.x/Startup.cs" 
name="Main" highlight-lines="2" data-bi-name="code-block">
CODE RENDERED HERE
</code>
```

mikaelm12 commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -16,127 +16,173 @@ uid: fundamentals/url-rewriting
+
+# [ASP.NET Core 2.x](#tab/aspnetcore2x)
+
+[!code-csharp[Main](url-rewriting/samples/2.x/Program.cs?name=snippet1)]
```
Here too

mikaelm12 commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -16,127 +16,173 @@ uid: fundamentals/url-rewriting
+
+# [ASP.NET Core 2.x](#tab/aspnetcore2x)
+
+[!code-csharp[Main](url-rewriting/samples/2.x/Program.cs?name=snippet1&highlight=5)]
```
Here and more below

mikaelm12 commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -16,127 +16,173 @@ uid: fundamentals/url-rewriting
+
+---
+
+In a browser with developer tools enabled, make a request to the sample app with the path `/redirect-rule/1234/5678`. The regex matches the request path on `redirect-rule/(.*)`, and the path is replaced with `/redirected/1234/5678`. The redirect URL is sent back to the client with a 302 (Found) status code. The browser makes a new request at the redirect URL, which appears in the browser's addres ...
```
Tempted also to put a note on pointing out that for a redirect the middleware rules will be evaluated again and emphasizing that  to help new users avoid hitting infinite redirects

mikaelm12 commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -175,30 +221,65 @@ The middleware supports the following Apache mod_rewrite server variables:
+* Outbound Rules
+* Custom Server Variables
+* Wildcards
+* Action:CustomResponse
```
We actually have support for the CustomResponse action now. https://github.com/aspnet/BasicMiddleware/pull/200

guardrex commented:
@mikaelm12 Got it.

* The first one that you called out for emphasis, I put into italics.
* The second one seemed important enough to "warn" them off of creating accidental infinite loops.

mikaelm12 commented on aspnetcore/fundamentals/url-rewriting.md:
```diff
@@ -16,127 +16,168 @@ uid: fundamentals/url-rewriting
+In a browser with developer tools enabled, make a request to the sample app with the path `/redirect-rule/1234/5678`. The regex matches the request path on `redirect-rule/(.*)`, and the path is replaced with `/redirected/1234/5678`. The redirect URL is sent back to the client with a 302 (Found) status code. The browser makes a new request at the redirect URL, which appears in the browser's addres ...
+
+> [!WARNING]
+> Be cautious when establishing your redirect rules. Your redirect rules are evaluated on each request to the app, including after a redirect. It's easy to accidently create a loop of infinite redirects.
```
Yes. This is great 👍

guardrex replied:
Cool. I like being very explicit about it ... "loop of infinite redirects" 😨 Yikes! lol

mikaelm12 replied:
It's redirects all the way down lol
